File: backend/api/routes/websocket.py
"""WebSocket endpoint for real-time telemetry."""

import asyncio
import json
from typing import Optional

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from backend.config import check_config_changed, get_config
from backend.simulation.engine import SimulationEngine


router = APIRouter()

# Global simulation engine instance
_engine: Optional[SimulationEngine] = None

# Config check interval (seconds)
CONFIG_CHECK_INTERVAL = 1.0


def get_engine() -> SimulationEngine:
    """Get or create simulation engine instance."""
    global _engine
    if _engine is None:
        _engine = SimulationEngine()
    return _engine


def reset_engine() -> SimulationEngine:
    """Reset engine with new config."""
    global _engine
    _engine = SimulationEngine(config=get_config())
    return _engine


class ConnectionManager:
    """Manages WebSocket connections."""

    def __init__(self):
        self.active_connections: list[WebSocket] = []

    async def connect(self, websocket: WebSocket) -> None:
        """Accept a new WebSocket connection."""
        await websocket.accept()
        self.active_connections.append(websocket)

    def disconnect(self, websocket: WebSocket) -> None:
        """Remove a WebSocket connection."""
        if websocket in self.active_connections:
            self.active_connections.remove(websocket)

    async def broadcast(self, message: dict) -> None:
        """Broadcast message to all connected clients."""
        disconnected = []
        for connection in self.active_connections:
            try:
                await connection.send_json(message)
            except Exception:
                disconnected.append(connection)

        # Clean up disconnected clients
        for conn in disconnected:
            self.disconnect(conn)


manager = ConnectionManager()


@router.websocket("/ws/telemetry")
async def telemetry_websocket(websocket: WebSocket):
    """WebSocket endpoint for real-time telemetry.

    Sends telemetry data at a configurable rate.
    Receives control commands from client.
    """
    await manager.connect(websocket)
    engine = get_engine()

    # Telemetry rate (Hz)
    telemetry_rate = 10.0
    telemetry_interval = 1.0 / telemetry_rate

    # Create separate tasks for sending and receiving
    send_task = asyncio.create_task(
        send_telemetry_loop(websocket, engine, telemetry_interval)
    )
    receive_task = asyncio.create_task(
        receive_message_loop(websocket, engine)
    )

    try:
        # Wait for either task to complete (usually due to disconnect)
        done, pending = await asyncio.wait(
            [send_task, receive_task],
            return_when=asyncio.FIRST_COMPLETED,
        )

        # Cancel pending tasks
        for task in pending:
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                pass

    except Exception:
        pass
    finally:
        # Ensure both tasks are cancelled
        send_task.cancel()
        receive_task.cancel()
        manager.disconnect(websocket)


def _step_and_get_telemetry(engine: SimulationEngine) -> dict:
    """Step simulation and get telemetry in a single call.

    This runs in thread pool to avoid blocking.
    """
    engine.step()
    telemetry = engine.get_telemetry()
    telemetry["type"] = "telemetry"
    return telemetry


async def send_telemetry_loop(
    websocket: WebSocket,
    engine: SimulationEngine,
    interval: float,
) -> None:
    """Background task to send telemetry at regular intervals.

    Uses asyncio.to_thread() to run simulation step in thread pool,
    preventing blocking of the async event loop during heavy computation.
    Also checks for config file changes periodically.
    """
    loop = asyncio.get_event_loop()
    last_config_check = loop.time()

    try:
        while True:
            loop_start = loop.time()

            # Check for config changes periodically (also in thread pool)
            if loop_start - last_config_check >= CONFIG_CHECK_INTERVAL:
                last_config_check = loop_start
                config_changed = await asyncio.to_thread(check_config_changed)
                if config_changed:
                    # Config changed - reset engine and notify client
                    engine = reset_engine()
                    await websocket.send_json({
                        "type": "config_reload",
                        "message": "Configuration reloaded, simulation reset",
                    })

            # Step simulation and get telemetry in thread pool
            telemetry = await asyncio.to_thread(_step_and_get_telemetry, engine)
            await websocket.send_json(telemetry)

            # Calculate remaining time to maintain consistent interval
            elapsed = loop.time() - loop_start
            sleep_time = max(0, interval - elapsed)
            if sleep_time > 0:
                await asyncio.sleep(sleep_time)
    except Exception as e:
        import logging
        logging.error(f"Telemetry loop error: {e}", exc_info=True)


async def receive_message_loop(
    websocket: WebSocket,
    engine: SimulationEngine,
) -> None:
    """Background task to receive and handle messages."""
    try:
        while True:
            data = await websocket.receive_text()
            await handle_message(data, engine, websocket)
    except WebSocketDisconnect:
        pass
    except Exception as e:
        import logging
        logging.error(f"Receive loop error: {e}", exc_info=True)


async def handle_message(
    data: str,
    engine: SimulationEngine,
    websocket: WebSocket,
) -> None:
    """Handle incoming WebSocket message.

    Args:
        data: JSON message string
        engine: Simulation engine instance
        websocket: WebSocket connection
    """
    try:
        message = json.loads(data)
        msg_type = message.get("type")

        if msg_type == "command":
            await handle_command(message, engine, websocket)
        elif msg_type == "mode":
            await handle_mode_change(message, engine, websocket)
        elif msg_type == "config":
            await handle_config(message, engine, websocket)
        elif msg_type == "timeline":
            await handle_timeline(message, engine, websocket)
        else:
            await send_error(websocket, f"Unknown message type: {msg_type}")

    except json.JSONDecodeError:
        await send_error(websocket, "Invalid JSON")
    except Exception as e:
        await send_error(websocket, str(e))


async def handle_command(
    message: dict,
    engine: SimulationEngine,
    websocket: WebSocket,
) -> None:
    """Handle control commands (START, STOP, PAUSE, RESET)."""
    command = message.get("command")

    if command == "START":
        engine.start()
    elif command == "STOP":
        engine.stop()
    elif command == "PAUSE":
        engine.pause()
    elif command == "RESET":
        engine.reset()
    else:
        await send_error(websocket, f"Unknown command: {command}")
        return

    await send_status(websocket, engine)


async def handle_mode_change(
    message: dict,
    engine: SimulationEngine,
    websocket: WebSocket,
) -> None:
    """Handle control mode change."""
    mode = message.get("mode")
    params = message.get("params", {})

    if mode in ["IDLE", "DETUMBLING", "POINTING", "UNLOADING"]:
        engine.set_control_mode(mode)

        # Handle pointing mode configuration
        if "pointingMode" in params:
            pointing_mode = params["pointingMode"]
            if pointing_mode in ["MANUAL", "SUN", "NADIR", "GROUND_STATION", "IMAGING_TARGET"]:
                engine.set_pointing_mode(pointing_mode)

        # Handle target quaternion for manual pointing mode
        if "targetQuaternion" in params:
            import numpy as np
            target = np.array(params["targetQuaternion"])
            engine.set_target_attitude(target)

        # Handle imaging target
        if "imagingTarget" in params:
            target = params["imagingTarget"]
            engine.set_imaging_target(
                lat_deg=target.get("latitude", 0),
                lon_deg=target.get("longitude", 0),
                alt_m=target.get("altitude", 0),
            )
    else:
        await send_error(websocket, f"Unknown mode: {mode}")
        return

    await send_status(websocket, engine)


async def handle_config(
    message: dict,
    engine: SimulationEngine,
    websocket: WebSocket,
) -> None:
    """Handle configuration changes."""
    if "timeWarp" in message:
        try:
            engine.set_time_warp(message["timeWarp"])
        except ValueError as e:
            await send_error(websocket, str(e))
            return

    await send_status(websocket, engine)


async def handle_timeline(
    message: dict,
    engine: SimulationEngine,
    websocket: WebSocket,
) -> None:
    """Handle timeline-related messages.

    Message formats:
        Add action:
            {"type": "timeline", "action": "add", "time": 1234.5,
             "actionType": "control_mode", "params": {"mode": "POINTING"}}
        Remove action:
            {"type": "timeline", "action": "remove", "actionId": "uuid"}
        Refresh contact:
            {"type": "timeline", "action": "refresh_contact"}
    """
    action = message.get("action")

    if action == "add":
        try:
            result = engine.add_timeline_action(
                time=message["time"],
                action_type=message["actionType"],
                params=message.get("params", {}),
            )
            await websocket.send_json({
                "type": "timeline_event",
                "event": "action_added",
                "action": result,
            })
        except ValueError as e:
            await send_error(websocket, str(e))
        except KeyError as e:
            await send_error(websocket, f"Missing required field: {e}")

    elif action == "remove":
        action_id = message.get("actionId")
        if not action_id:
            await send_error(websocket, "Missing actionId")
            return

        if engine.remove_timeline_action(action_id):
            await websocket.send_json({
                "type": "timeline_event",
                "event": "action_removed",
                "actionId": action_id,
            })
        else:
            await send_error(websocket, "Action not found")

    elif action == "refresh_contact":
        contact = engine.refresh_contact_prediction()
        await websocket.send_json({
            "type": "timeline_event",
            "event": "contact_refreshed",
            "nextContact": contact,
        })

    elif action == "imaging_preset":
        offset = message.get("offsetSeconds", 300.0)
        schedule = message.get("scheduleAction", False)
        result = engine.set_imaging_preset(
            offset_seconds=offset,
            schedule_action=schedule,
        )
        if result:
            await websocket.send_json({
                "type": "timeline_event",
                "event": "imaging_preset_set",
                "preset": result,
            })
        else:
            await send_error(websocket, "No contact predicted for imaging preset")

    else:
        await send_error(websocket, f"Unknown timeline action: {action}")


async def send_status(websocket: WebSocket, engine: SimulationEngine) -> None:
    """Send status update to client."""
    status = {
        "type": "status",
        "state": engine.state.name,
        "simTime": engine.sim_time,
        "timeWarp": engine.time_warp,
    }
    await websocket.send_json(status)


async def send_error(websocket: WebSocket, message: str) -> None:
    """Send error message to client."""
    error = {
        "type": "error",
        "message": message,
    }
    await websocket.send_json(error)

File: frontend/src/services/websocket.ts
/**
 * WebSocket service for real-time telemetry communication.
 */

import type {
  Telemetry,
  ClientMessage,
  WebSocketMessage,
  ControlMode,
  PointingMode,
  ImagingTarget,
  TimelineActionType,
  PointingConfig,
} from '../types/telemetry';

type MessageHandler = (message: WebSocketMessage) => void;
type ConnectionHandler = () => void;
type ErrorHandler = (error: Event) => void;

export class TelemetryWebSocket {
  private ws: WebSocket | null = null;
  private url: string;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private messageHandlers: Set<MessageHandler> = new Set();
  private connectHandlers: Set<ConnectionHandler> = new Set();
  private disconnectHandlers: Set<ConnectionHandler> = new Set();
  private errorHandlers: Set<ErrorHandler> = new Set();

  constructor(url: string = 'ws://localhost:8000/ws/telemetry') {
    this.url = url;
  }

  /**
   * Connect to the WebSocket server.
   */
  connect(): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      return;
    }

    try {
      this.ws = new WebSocket(this.url);

      this.ws.onopen = () => {
        console.log('WebSocket connected');
        this.reconnectAttempts = 0;
        this.connectHandlers.forEach(handler => handler());
      };

      this.ws.onclose = () => {
        console.log('WebSocket disconnected');
        this.disconnectHandlers.forEach(handler => handler());
        this.attemptReconnect();
      };

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error);
        this.errorHandlers.forEach(handler => handler(error));
      };

      this.ws.onmessage = (event) => {
        try {
          const message: WebSocketMessage = JSON.parse(event.data);
          this.messageHandlers.forEach(handler => handler(message));
        } catch (e) {
          console.error('Failed to parse WebSocket message:', e);
        }
      };
    } catch (error) {
      console.error('Failed to create WebSocket:', error);
    }
  }

  /**
   * Disconnect from the WebSocket server.
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Send a message to the server.
   */
  send(message: ClientMessage): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message));
    } else {
      console.warn('WebSocket not connected');
    }
  }

  /**
   * Send simulation control command.
   */
  sendCommand(command: 'START' | 'STOP' | 'PAUSE' | 'RESET'): void {
    this.send({ type: 'command', command });
  }

  /**
   * Set control mode with optional pointing configuration.
   */
  setControlMode(
    mode: ControlMode,
    options?: {
      pointingMode?: PointingMode;
      targetQuaternion?: [number, number, number, number];
      imagingTarget?: ImagingTarget;
    }
  ): void {
    const message: ClientMessage = {
      type: 'mode',
      mode,
      params: options,
    };
    this.send(message);
  }

  /**
   * Set time warp.
   */
  setTimeWarp(timeWarp: number): void {
    this.send({ type: 'config', timeWarp });
  }

  /**
   * Set detailed pointing configuration with main/sub axis.
   */
  setPointingConfig(config: PointingConfig): void {
    this.send({
      type: 'pointing_config',
      mainTarget: config.mainTarget,
      mainBodyAxis: config.mainBodyAxis,
      subTarget: config.subTarget,
      subBodyAxis: config.subBodyAxis,
    });
  }

  // ==================== Timeline Methods ====================

  /**
   * Add a scheduled action to the timeline.
   */
  addTimelineAction(
    time: number,
    actionType: TimelineActionType,
    params: Record<string, unknown>
  ): void {
    this.send({
      type: 'timeline',
      action: 'add',
      time,
      actionType,
      params,
    });
  }

  /**
   * Remove a scheduled action from the timeline.
   */
  removeTimelineAction(actionId: string): void {
    this.send({
      type: 'timeline',
      action: 'remove',
      actionId,
    });
  }

  /**
   * Force refresh contact prediction.
   */
  refreshContactPrediction(): void {
    this.send({
      type: 'timeline',
      action: 'refresh_contact',
    });
  }

  /**
   * Set imaging target from contact + offset.
   */
  setImagingPreset(offsetSeconds: number = 300, scheduleAction: boolean = false): void {
    this.send({
      type: 'timeline',
      action: 'imaging_preset',
      offsetSeconds,
      scheduleAction,
    });
  }

  /**
   * Subscribe to messages.
   */
  onMessage(handler: MessageHandler): () => void {
    this.messageHandlers.add(handler);
    return () => this.messageHandlers.delete(handler);
  }

  /**
   * Subscribe to telemetry updates only.
   */
  onTelemetry(handler: (telemetry: Telemetry) => void): () => void {
    const wrappedHandler: MessageHandler = (message) => {
      if (message.type === 'telemetry') {
        handler(message);
      }
    };
    return this.onMessage(wrappedHandler);
  }

  /**
   * Subscribe to connection events.
   */
  onConnect(handler: ConnectionHandler): () => void {
    this.connectHandlers.add(handler);
    return () => this.connectHandlers.delete(handler);
  }

  /**
   * Subscribe to disconnection events.
   */
  onDisconnect(handler: ConnectionHandler): () => void {
    this.disconnectHandlers.add(handler);
    return () => this.disconnectHandlers.delete(handler);
  }

  /**
   * Subscribe to error events.
   */
  onError(handler: ErrorHandler): () => void {
    this.errorHandlers.add(handler);
    return () => this.errorHandlers.delete(handler);
  }

  /**
   * Check if connected.
   */
  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  private attemptReconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.log('Max reconnect attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

    console.log(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`);

    setTimeout(() => {
      this.connect();
    }, delay);
  }
}

// Singleton instance
export const telemetryWS = new TelemetryWebSocket();
